btk/sampling_functions.py

```diff
@@ -429,3 +431,105 @@ def __call__(self, table: Table):
         blend_table["dec"] *= 3600
 
         return blend_table
+
+
+class FriendsOfFriendsSampling(SamplingFunction):
+    """Randomly selects galaxies using Friends Of Friends clustering algorithm.
+
+    Friends of friends clustering algorithm assigns a group of object same index if one can
+    each member of the group is within link_distance of any other member in the group.
+    This sampling function explicitly uses the spatial information in the input catalog to
+    generate scenes of galaxies. However, blends might not always be returned as a result.
+    """
+
+    def __init__(
+        self,
+        max_number: int = 10,
+        min_number: int = 2,
+        link_distance: int = 2.5,
+        stamp_size: float = 24.0,
+        seed: int = DEFAULT_SEED,
+        min_mag: float = -np.inf,
+        max_mag: float = 25.3,
+        mag_name: str = "i_ab",
+    ):
+        """Initializes the FriendsOfFriendsSampling sampling function.
+
+        Args:
+            max_number: Defined in parent class
+            min_number: Defined in parent class
+            link_distance: Minimum linkage distance to form a group (arcsec).
+            stamp_size: Size of the desired stamp (arcsec).
+            seed: Seed to initialize randomness for reproducibility.
+            min_mag: Minimum magnitude allowed in samples
+            max_mag: Maximum magnitude allowed in samples.
+            mag_name: Name of the magnitude column in the catalog.
+        """
+        super().__init__(max_number=max_number, min_number=min_number, seed=seed)
+        self.stamp_size = stamp_size
+        self.link_distance = link_distance
+        self.max_number = max_number
+        self.min_number = min_number
+        self.max_mag = max_mag
+        self.min_mag = min_mag
+        self.mag_name = mag_name
+
+    @staticmethod
+    def _arcsec2dist(sep, r=1.0):
```

Comment:
Why do you need the optional argumentn `r`? and what does it represent?

Reply:
so to convert the distance between the two objects on the sky to ecludian distance you generally need three numbers delta r = r2 - r1, delta phi = phi2 - phi1, delta theta = theta2 - theta1. but we don't know distance to the objects necessarily, so the assumption for calculating distances is that they are all at the same distance from the origin (same redshift). r = 1 is really just a normalization factor, which tells you how how ecludian distances scale with delta theta and delta phi.

Reply:
setting r = const for all objects seems like a rough assumption at first, but that's what we mean by a blend. we say that two galaxies are blended if they are close in delta phi and delta theta, and it doesn't matter that they are at a different redshifts

Reply:
that makes sense!